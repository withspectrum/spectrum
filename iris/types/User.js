const User = /* GraphQL */ `
	type UserCommunitiesConnection {
		pageInfo: PageInfo!
		edges: [UserCommunityEdge!]
	}

	type UserCommunityEdge {
		node: Community!
	}

	type UserChannelsConnection {
		pageInfo: PageInfo!
		edges: [UserChannelEdge!]
	}

	type UserChannelEdge {
		node: Channel!
	}

	type UserDirectMessageThreadsConnection {
		pageInfo: PageInfo!
		edges: [DirectMessageThreadEdge]
	}

	type DirectMessageThreadEdge {
		cursor: String!
		node: DirectMessageThread!
	}

	type UserThreadsConnection {
		pageInfo: PageInfo!
		edges: [UserThreadEdge!]
	}

	type UserThreadEdge {
		cursor: String!
		node: Thread!
	}

	type EverythingThreadsConnection {
		pageInfo: PageInfo!
		edges: [EverythingThreadEdge!]
	}

	type EverythingThreadEdge {
		cursor: String!
		node: Thread!
	}

	type UserMetaData {
		threads: Int
	}

	type NotificationKindSettings {
		email: Boolean
	}

	type NotificationSettingsType {
		newMessageInThreads: NotificationKindSettings
		newThreadCreated: NotificationKindSettings
		weeklyDigest: NotificationKindSettings
		dailyDigest: NotificationKindSettings
	}

	type UserNotificationsSettings {
		types: NotificationSettingsType
	}

	type UserSettings {
		notifications: UserNotificationsSettings
	}

	type User {
		id: ID!
		name: String
		description: String
		website: String
		username: String
		profilePhoto: String
		coverPhoto: String
		email: String
		providerId: String
		createdAt: Date!
		lastSeen: Date!
		isOnline: Boolean
		timezone: Int
		totalReputation: Int

		# non-schema fields
		threadCount: Int
		isAdmin: Boolean!
		isPro: Boolean!
		communityConnection: UserCommunitiesConnection!
		channelConnection: UserChannelsConnection!
		directMessageThreadsConnection: UserDirectMessageThreadsConnection!
		threadConnection(first: Int = 20, after: String): UserThreadsConnection!
		everything(first: Int = 20, after: String): EverythingThreadsConnection!
		recurringPayments: [RecurringPayment]
		invoices: [Invoice]
		settings: UserSettings
		contextPermissions: ContextPermissions
	}

	extend type Query {
		user(id: ID, username: String): User
		currentUser: User
		searchUsers(string: String): [User]
	}

	input EditUserInput {
		file: File
		coverFile: File
		name: String
		description: String
		website: String
		username: String
		timezone: Int
	}

	input UpgradeToProInput {
		plan: String!
		token: String!
	}

	input ToggleNotificationSettingsInput {
		deliveryMethod: String!
		notificationType: String!
	}

	input WebPushSubscriptionKeys {
		p256dh: String!
		auth: String!
	}

	input WebPushSubscription {
		endpoint: String!
		keys: WebPushSubscriptionKeys!
	}

	extend type Mutation {
		editUser(input: EditUserInput!): User
		upgradeToPro(input: UpgradeToProInput!): User
		downgradeFromPro: User
		toggleNotificationSettings(input: ToggleNotificationSettingsInput): User
		subscribeWebPush(subscription: WebPushSubscription!): Boolean
		unsubscribeWebPush(endpoint: String!): Boolean
	}
`;

module.exports = User;
