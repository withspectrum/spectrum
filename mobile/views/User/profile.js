// @flow
import React, { Component, Fragment } from 'react';
import { Text, View, StatusBar } from 'react-native';
import compose from 'recompose/compose';
import getUserThreadConnection from '../../../shared/graphql/queries/user/getUserThreadConnection';
import ThreadFeed from '../../components/ThreadFeed';
import type { GetUserType } from '../../../shared/graphql/queries/user/getUser';

import {
  Wrapper,
  CoverPhoto,
  CoverPhotoFill,
  CoverPhotoContainer,
  ProfilePhoto,
  ProfilePhotoContainer,
  ProfileDetailsContainer,
  Name,
  Username,
  Description,
  ThreadFeedTabContainer,
  ThreadFeedTab,
  TabLabel,
} from './style';

type Props = {
  isLoading: boolean,
  hasError: boolean,
  navigation: Object,
  data: {
    user?: GetUserType,
  },
};

type State = {
  feed: string,
};

const UserThreadFeed = compose(getUserThreadConnection)(ThreadFeed);

class User extends Component<Props, State> {
  state = { feed: 'participant' };

  componentDidUpdate() {
    const { data: { user }, navigation } = this.props;
    if (!user) return;
    const title = navigation.getParam('title');
    if (!title && user) return navigation.setParams({ title: user.name });
    if (title && title !== user.name)
      return navigation.setParams({ title: user.name });
  }

  toggleFeed = (feed: string) => this.setState({ feed });

  render() {
    const { data, isLoading, hasError, navigation } = this.props;
    const { feed } = this.state;

    if (data.user) {
      return (
        <Wrapper>
          <StatusBar barStyle="light-content" />
          <UserThreadFeed
            navigation={navigation}
            kind={this.state.feed}
            id={data.user.id}
            ListHeaderComponent={
              <Fragment>
                <CoverPhotoContainer>
                  {data.user.coverPhoto ? (
                    <CoverPhoto
                      resizeMode={'cover'}
                      source={{ uri: data.user.coverPhoto }}
                    />
                  ) : (
                    <CoverPhotoFill />
                  )}
                </CoverPhotoContainer>

                <ProfilePhotoContainer>
                  <ProfilePhoto source={{ uri: data.user.profilePhoto }} />
                </ProfilePhotoContainer>

                <ProfileDetailsContainer>
                  <Name>{data.user.name}</Name>
                  <Username>@{data.user.username}</Username>
                  <Description>{data.user.description}</Description>
                </ProfileDetailsContainer>

                <ThreadFeedTabContainer>
                  <ThreadFeedTab
                    onPress={() => this.toggleFeed('participant')}
                    isActive={feed === 'participant'}
                  >
                    <TabLabel isActive={feed === 'participant'}>
                      Replies
                    </TabLabel>
                  </ThreadFeedTab>
                  <ThreadFeedTab
                    onPress={() => this.toggleFeed('creator')}
                    isActive={feed === 'creator'}
                  >
                    <TabLabel isActive={feed === 'creator'}>Threads</TabLabel>
                  </ThreadFeedTab>
                </ThreadFeedTabContainer>
              </Fragment>
            }
          />
        </Wrapper>
      );
    }

    if (isLoading) {
      return (
        <Wrapper>
          <View testID="e2e-User">
            <Text>Loading...</Text>
          </View>
        </Wrapper>
      );
    }

    if (hasError) {
      return (
        <Wrapper>
          <View testID="e2e-User">
            <Text>Error!</Text>
          </View>
        </Wrapper>
      );
    }

    return null;
  }
}

export default User;
