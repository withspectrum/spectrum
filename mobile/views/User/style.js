// @flow
import styled from 'styled-components/native';
import { Dimensions } from 'react-native';
import TouchableOpacity from '../../components/TouchableOpacity';

const width = Dimensions.get('window').width;

export const Wrapper = styled.View`
  background-color: ${props => props.theme.bg.default};
  flex: 1;
  align-items: center;
  justify-content: center;
  align-self: stretch;
  width: ${width};
`;

export const CoverPhotoContainer = styled.View`
  width: ${width};
  height: 124px;
  display: flex;
  align-self: stretch;
`;

export const CoverPhoto = styled.Image`
  width: ${width};
  height: 124px;
`;

export const CoverPhotoFill = styled.View`
  width: ${width};
  height: 124px;
  background: ${props => props.theme.brand.alt};
`;

export const ProfilePhotoContainer = styled.View`
  width: ${width};
  display: flex;
  justify-content: flex-start;
  padding: 0 16px;
  position: relative;
  top: -32px;
  margin-bottom: -24px;
`;

export const ProfilePhoto = styled.Image`
  width: 72px;
  height: 72px;
  border-radius: 36px;
  border-width: 4px;
  border-color: ${props => props.theme.bg.default};
  shadow-color: ${props => props.theme.bg.default};
  shadow-opacity: 1;
  shadow-radius: 4;
`;

export const ProfileDetailsContainer = styled.View`
  display: flex;
  flex-direction: column;
  width: ${width};
  padding: 0 24px;
`;

export const Name = styled.Text`
  font-weight: 800;
  color: ${props => props.theme.text.default};
  font-size: 20px;
`;

export const Username = styled.Text`
  font-weight: 500;
  color: ${props => props.theme.text.alt};
  font-size: 16px;
  margin-top: 4px;
`;

export const Description = styled.Text`
  font-weight: 400;
  color: ${props => props.theme.text.default};
  font-size: 16px;
  margin-top: 8px;
`;

export const ThreadFeedTabContainer = styled.View`
  display: flex;
  flex-direction: row;
  width: ${width};
  margin-top: 16px;
  border-bottom-color: ${props => props.theme.bg.border};
  border-bottom-width: 1px;
`;

export const ThreadFeedTab = styled(TouchableOpacity)`
  padding: 8px 16px;
  display: flex;
  flex: 1;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  border-bottom-color: ${props =>
    props.isActive ? props.theme.text.default : 'transparent'};
  border-bottom-width: 1px;
`;
export const TabLabel = styled.Text`
  color: ${props =>
    props.isActive ? props.theme.text.default : props.theme.text.alt};
  font-size: 16px;
  font-weight: ${props => (props.isActive ? '500' : '400')};
`;
