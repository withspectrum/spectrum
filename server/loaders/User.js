// @flow
import { getUsers, getUsersStoryCount } from '../models/user';
import createLoader from './create-loader';
import type { Loader } from './types';

export const __createUserLoader = () =>
  createLoader(users => getUsers(users), 'uid');

export const __createUserStoryCountLoader = () =>
  createLoader(users => getUsersStoryCount(users), 'uid');

export default () => {
  throw new Error(
    '⚠️ Do not import loaders directly, get them from the GraphQL context instead! ⚠️'
  );
};
