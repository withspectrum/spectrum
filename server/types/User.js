const User = /* GraphQL */ `
	type UserCommunitiesConnection {
		pageInfo: PageInfo!
		edges: [UserCommunityEdge!]
	}

	type UserCommunityEdge {
		node: Community!
	}

	type UserFrequenciesConnection {
		pageInfo: PageInfo!
		edges: [UserFrequencyEdge!]
	}

	type UserFrequencyEdge {
		node: Frequency!
	}

	type UserDirectMessageGroupsConnection {
		pageInfo: PageInfo!
		edges: [DirectMessageGroupEdge]
	}

	type DirectMessageGroupEdge {
		cursor: String!
		node: DirectMessageGroup!
	}

	type UserStoriesConnection {
		pageInfo: PageInfo!
		edges: [UserStoryEdge!]
	}

	type UserStoryEdge {
		cursor: String!
		node: Story!
	}

	type EverythingStoriesConnection {
		pageInfo: PageInfo!
		edges: [EverythingStoryEdge!]
	}

	type EverythingStoryEdge {
		cursor: String!
		node: Story!
	}

	type UserMetaData {
		stories: Int
	}

	type User {
		uid: ID!
		createdAt: Date!
		lastSeen: Date!
		photoURL: String
		displayName: String
		username: String
		email: String
		# subscriptions: [Subscription!]
		communityConnection: UserCommunitiesConnection!
		frequencyConnection: UserFrequenciesConnection!
		directMessageGroupsConnection: UserDirectMessageGroupsConnection!
		storyConnection(first: Int = 10, after: String): UserStoriesConnection!
		everything(first: Int = 10, after: String): EverythingStoriesConnection!
		metaData: UserMetaData!
	}

	extend type Query {
		user(uid: ID, username: String): User
		currentUser: User
	}
`;

module.exports = User;
