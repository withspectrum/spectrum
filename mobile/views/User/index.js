// @flow
import * as React from 'react';
import { Text, View, FlatList, Button, ScrollView } from 'react-native';
import compose from 'recompose/compose';
import { getUserById } from '../../../shared/graphql/queries/user/getUser';
import getUserThreadConnection from '../../../shared/graphql/queries/user/getUserThreadConnection';
import ViewNetworkHandler from '../../components/ViewNetworkHandler';
import withSafeView from '../../components/SafeAreaView';
import ThreadFeed from '../../components/ThreadFeed';

import { Wrapper } from './style';

type UserType = {
  id: string,
  name: string,
  username: string,
  threadConnection: {
    pageInfo: {
      hasNextPage: boolean,
    },
    edges: {
      node: {
        id: string,
      },
    },
  },
};

type Props = {
  isLoading: boolean,
  hasError: boolean,
  navigation: Object,
  data: {
    user?: UserType,
  },
};

type State = {
  feed: string,
};

const UserThreadFeed = compose(getUserThreadConnection)(ThreadFeed);

class User extends React.Component<Props, State> {
  state = { feed: 'participant' };

  componentDidUpdate() {
    const { data: { user }, navigation } = this.props;
    if (!user || navigation.state.params.title) return;
    navigation.setParams({ title: `${user.name} (@${user.username})` });
  }

  toggleFeed = (feed: string) => this.setState({ feed });

  render() {
    const { data, isLoading, hasError, navigation } = this.props;
    const { feed } = this.state;

    if (data.user) {
      return (
        <Wrapper>
          <Button
            title={'View active conversations'}
            onPress={() => this.toggleFeed('participant')}
          />

          <Button
            title={'View created conversations'}
            onPress={() => this.toggleFeed('creator')}
          />

          <Text>Viewing {this.state.feed} thread feed</Text>

          <UserThreadFeed
            navigation={navigation}
            kind={this.state.feed}
            id={data.user.id}
          />
        </Wrapper>
      );
    }

    if (isLoading) {
      return (
        <Wrapper>
          <View testID="e2e-User">
            <Text>Loading...</Text>
          </View>
        </Wrapper>
      );
    }

    if (hasError) {
      return (
        <Wrapper>
          <View testID="e2e-User">
            <Text>Error!</Text>
          </View>
        </Wrapper>
      );
    }

    return null;
  }
}

export default compose(withSafeView, getUserById, ViewNetworkHandler)(User);
