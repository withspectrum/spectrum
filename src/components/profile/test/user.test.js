// @flow
import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import User from '../user';

const data = {
  title: 'Test user',
  subtitle: 'Whats up',
  description: 'Testing the user profile',
  photoURL: 'img.com/asdf.jpg',
  id: 'asdf-123',
};

it('should render with no size', () => {
  const wrapper = shallow(<User data={data} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

it('should render with a mini size', () => {
  const wrapper = shallow(<User size="mini" data={data} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

it('should render with a small size', () => {
  const wrapper = shallow(<User size="small" data={data} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

it('should render with a medium size', () => {
  const wrapper = shallow(<User size="medium" data={data} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

it('should render with a large size', () => {
  const wrapper = shallow(<User size="large" data={data} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

it('should render with a full size', () => {
  const wrapper = shallow(<User size="full" data={data} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});
