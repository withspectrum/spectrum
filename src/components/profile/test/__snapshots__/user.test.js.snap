exports[`test should render with a full size 1`] = `
<pure(CardPure)
  data={
    Object {
      "description": "Testing the user profile",
      "id": "asdf-123",
      "photoURL": "img.com/asdf.jpg",
      "subtitle": "Whats up",
      "title": "Test user",
    }
  }
  size="full">
  <Styled()
    alignItems="center"
    justifyContent="flex-start">
    <pure(AvatarPure)
      margin="0 12px 0 0"
      radius={4}
      size={40}
      src="img.com/asdf.jpg" />
    <Styled()
      direction="column"
      justifyContent="center">
      <styled.h3>
        Test user
      </styled.h3>
      <styled.h4>
        Whats up
      </styled.h4>
    </Styled()>
  </Styled()>
  <styled.p>
    Testing the user profile
  </styled.p>
  <Styled()>
    <Styled()>
      Message
    </Styled()>
    <Styled(pure(ButtonPure))>
      Follow
    </Styled(pure(ButtonPure))>
  </Styled()>
  <Apollo(branch(pure(MetaDataPure)))
    id="asdf-123"
    type="user" />
</pure(CardPure)>
`;

exports[`test should render with a large size 1`] = `
<pure(CardPure)
  data={
    Object {
      "description": "Testing the user profile",
      "id": "asdf-123",
      "photoURL": "img.com/asdf.jpg",
      "subtitle": "Whats up",
      "title": "Test user",
    }
  }
  size="large">
  <Styled()
    alignItems="center"
    justifyContent="flex-start">
    <pure(AvatarPure)
      margin="0 12px 0 0"
      radius={4}
      size={40}
      src="img.com/asdf.jpg" />
    <Styled()
      direction="column"
      justifyContent="center">
      <styled.h3>
        Test user
      </styled.h3>
      <styled.h4>
        Whats up
      </styled.h4>
    </Styled()>
  </Styled()>
  <styled.p>
    Testing the user profile
  </styled.p>
  <Styled()>
    <Styled()>
      Message
    </Styled()>
  </Styled()>
  <Apollo(branch(pure(MetaDataPure)))
    id="asdf-123"
    type="user" />
</pure(CardPure)>
`;

exports[`test should render with a medium size 1`] = `
<pure(CardPure)
  data={
    Object {
      "description": "Testing the user profile",
      "id": "asdf-123",
      "photoURL": "img.com/asdf.jpg",
      "subtitle": "Whats up",
      "title": "Test user",
    }
  }
  size="medium">
  <Styled()
    alignItems="center"
    justifyContent="flex-start">
    <pure(AvatarPure)
      margin="0 12px 0 0"
      radius={4}
      size={40}
      src="img.com/asdf.jpg" />
    <Styled()
      direction="column"
      justifyContent="center">
      <styled.h3>
        Test user
      </styled.h3>
      <styled.h4>
        Whats up
      </styled.h4>
    </Styled()>
  </Styled()>
  <styled.p>
    Testing the user profile
  </styled.p>
  <Styled()>
    <Styled()>
      Message
    </Styled()>
  </Styled()>
</pure(CardPure)>
`;

exports[`test should render with a mini size 1`] = `
<pure(CardPure)
  data={
    Object {
      "description": "Testing the user profile",
      "id": "asdf-123",
      "photoURL": "img.com/asdf.jpg",
      "subtitle": "Whats up",
      "title": "Test user",
    }
  }
  size="mini">
  <Styled()
    alignItems="center"
    justifyContent="flex-start">
    <pure(AvatarPure)
      margin="0 12px 0 0"
      radius={4}
      size={40}
      src="img.com/asdf.jpg" />
    <Styled()
      direction="column"
      justifyContent="center">
      <styled.h3>
        Test user
      </styled.h3>
      <styled.h4>
        Whats up
      </styled.h4>
    </Styled()>
  </Styled()>
</pure(CardPure)>
`;

exports[`test should render with a small size 1`] = `
<pure(CardPure)
  data={
    Object {
      "description": "Testing the user profile",
      "id": "asdf-123",
      "photoURL": "img.com/asdf.jpg",
      "subtitle": "Whats up",
      "title": "Test user",
    }
  }
  size="small">
  <Styled()
    alignItems="center"
    justifyContent="flex-start">
    <pure(AvatarPure)
      margin="0 12px 0 0"
      radius={4}
      size={40}
      src="img.com/asdf.jpg" />
    <Styled()
      direction="column"
      justifyContent="center">
      <styled.h3>
        Test user
      </styled.h3>
      <styled.h4>
        Whats up
      </styled.h4>
    </Styled()>
  </Styled()>
  <Styled()>
    <Styled()>
      Message
    </Styled()>
  </Styled()>
</pure(CardPure)>
`;

exports[`test should render with no size 1`] = `
<pure(CardPure)
  data={
    Object {
      "description": "Testing the user profile",
      "id": "asdf-123",
      "photoURL": "img.com/asdf.jpg",
      "subtitle": "Whats up",
      "title": "Test user",
    }
  }>
  <Styled()
    alignItems="center"
    justifyContent="flex-start">
    <pure(AvatarPure)
      margin="0 12px 0 0"
      radius={4}
      size={40}
      src="img.com/asdf.jpg" />
    <Styled()
      direction="column"
      justifyContent="center">
      <styled.h3>
        Test user
      </styled.h3>
      <styled.h4>
        Whats up
      </styled.h4>
    </Styled()>
  </Styled()>
</pure(CardPure)>
`;
